sklearn_numba_dpex/kmeans/kernels/utils.py

```diff
@@ -249,3 +250,84 @@ def reduce_centroid_data(
                 empty_clusters_list[current_n_empty_clusters] = cluster_idx
 
     return reduce_centroid_data[global_size, work_group_size]
+
+
+@lru_cache
+def make_is_same_clustering_kernel(n_samples, n_clusters, work_group_size, device):
+    # TODO: are there possible optimizations for this kernel ?
+    # - fusing the two kernels (requires a lock ?)
+    # - early stop
+
+    def is_same_clustering(labels1, labels2):
+        mapping = dpt.empty(sh=(n_clusters,), dtype=np.int32, device=device)
+        result = dpt.asarray([1], dtype=np.int32, device=device)
+        _build_mapping[global_size, work_group_size](labels1, labels2, mapping)
+        _is_same_clustering[global_size, work_group_size](
+            labels1, labels2, mapping, result
+        )
+        return bool(result[0])
+
+    @dpex.kernel
+    # fmt: off
+    def _build_mapping(
+        labels1,            # IN      (n_samples,)
+        labels2,            # IN      (n_samples,)
+        mapping,            # BUFFER  (n_clusters,)
+    ):
+        # fmt: on
+        sample_idx = dpex.get_global_id(zero_idx)
+        if sample_idx >= n_samples:
+            return
+
+        label1 = labels1[sample_idx]
+        label2 = labels2[sample_idx]
+        mapping[label1] = label2
+
+    @dpex.kernel
+    # fmt: off
+    def _is_same_clustering(
+        labels1,
+        labels2,
+        mapping,
+        result
+    ):
+        # fmt: on
+        """`result` is expected to be an empty array with dtype int32 of size (1,)
+        initialized with value 1.
+        """
+        sample_idx = dpex.get_global_id(zero_idx)
+        if sample_idx >= n_samples:
```

Comment:
Seems better but note that this add a read operation and doesn't stop future tasks, it justs ensure that future tasks will exit early after the read. So on average it's better when clusterings are not the same but not when clusterings are the same. Still better since different clusterings should be the most common outcome.